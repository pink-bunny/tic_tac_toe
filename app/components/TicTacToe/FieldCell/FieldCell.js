import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const FieldCell = ({
  disabled, onHandleClick, className, visible, value
}) => (
  <button
    type="button"
    disabled={disabled}
    onClick={onHandleClick}
    className={classNames('ttt-field__item', className)}
  >
    {visible && value}
  </button>
);

FieldCell.propTypes = {
  disabled: PropTypes.bool,
  visible: PropTypes.bool,
  className: PropTypes.string,
  value: PropTypes.string,
  onHandleClick: PropTypes.func
};

export default FieldCell;
