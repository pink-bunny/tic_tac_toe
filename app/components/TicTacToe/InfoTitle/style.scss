.ttt-title {
  text-align: center;
  margin-top: 0;
  font-weight: 400;
}
.ttt-title--win {
  color: #41ADDD;
}
.ttt-title--draw {
  color: #dd4141;
}
.ttt-title__sign {
  font-weight: bold;
  text-transform: uppercase;
}
