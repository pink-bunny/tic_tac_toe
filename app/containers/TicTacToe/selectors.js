import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectTicTacToe = (state) => state.ticTacToe || initialState;

const selectTTTFieldLength = () => createSelector(
  selectTicTacToe,
  (ticTacToelState) => ticTacToelState.fieldLength
);

const selectTTTCurrentPlayer = () => createSelector(
  selectTicTacToe,
  (ticTacToelState) => ticTacToelState.currentPlayer
);

const selectTTTSetsPlayed = () => createSelector(
  selectTicTacToe,
  (ticTacToelState) => ticTacToelState.setsPlayed
);

const selectTTTPlayer1 = () => createSelector(
  selectTicTacToe,
  (ticTacToelState) => ticTacToelState.player1
);

const selectTTTPlayer2 = () => createSelector(
  selectTicTacToe,
  (ticTacToelState) => ticTacToelState.player2
);

export {
  selectTicTacToe,
  selectTTTFieldLength,
  selectTTTSetsPlayed,
  selectTTTPlayer1,
  selectTTTPlayer2,
  selectTTTCurrentPlayer
};
