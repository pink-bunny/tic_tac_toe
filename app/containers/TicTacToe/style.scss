#app {
  background-image: url("https://mdbootstrap.com/img/Photos/Horizontal/Nature/full%20page/img(11).jpg");
  background-size: cover;
}
.ttt {
  padding-top: 60px;
  width: 100%;
  max-width: 900px;
  margin: 0 auto;
}
.ttt-field-wrap {
  width: max-content;
  margin: 0 auto;
}
.ttt-field {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr 1fr;
  width: 300px;
  height: 300px;
}
.ttt-field__item {
  transition: all 0.2s;
  display: flex;
  align-items: center;
  justify-content: center;
  color: white;
  font-size: 24px;
  text-transform: uppercase;
  border: 1px solid rgba(255, 255, 255, 0.75);
  background-color: rgba(16, 42, 64, 0.75);
  cursor: pointer;
  &:focus {
    outline: none;
  }
  &:disabled {
    cursor: not-allowed;
  }
}
.ttt-field__item--win {
  background-color: rgba(65, 173, 221, 0.65);
  font-weight: bold;
}
