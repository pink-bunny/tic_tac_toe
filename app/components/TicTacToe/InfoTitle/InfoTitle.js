import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './style.scss';

const InfoTitle = ({ children, className }) => (
  <h2 className={classNames('ttt-title', className)}>
    {children}
  </h2>
);

InfoTitle.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node
};

export default InfoTitle;
