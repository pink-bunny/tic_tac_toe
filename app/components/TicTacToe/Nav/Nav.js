import React from 'react';
import PropTypes from 'prop-types';

import './style.scss';

const Nav = ({ onBack, onForward }) => (
  <div className="ttt-nav">
    <button
      type="button"
      className="ttt-nav__btn"
      onClick={onBack}
    >
       Step Back
    </button>
    <button
      type="button"
      className="ttt-nav__btn"
      onClick={onForward}
    >
      Step Forward
    </button>
  </div>
);

Nav.propTypes = {
  onBack: PropTypes.func,
  onForward: PropTypes.func
};

export default Nav;
