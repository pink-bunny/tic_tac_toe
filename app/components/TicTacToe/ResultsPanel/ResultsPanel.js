import React from 'react';
import PropTypes from 'prop-types';

import './style.scss';

const ResultsPanel = ({ setsPlayed, player1, player2 }) => (
  <div className="ttt-result">

    <p className="ttt-result__item">
      Sets Played:
      <strong>{` ${setsPlayed}`}</strong>
    </p>

    <p className="ttt-result__item">
      Player
      {` "${player1.value}" `}
      wins:
      <strong>{` ${player1.wins}`}</strong>
    </p>

    <p className="ttt-result__item">
      Player
      {` "${player2.value}" `}
      wins:
      <strong>{` ${player2.wins}`}</strong>
    </p>

  </div>
);

ResultsPanel.propTypes = {
  setsPlayed: PropTypes.number,
  player1: PropTypes.object,
  player2: PropTypes.object
};

export default ResultsPanel;
