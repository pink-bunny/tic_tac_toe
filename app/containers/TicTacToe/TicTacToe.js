import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import Nav from '../../components/TicTacToe/Nav';
import NewGameBtn from '../../components/TicTacToe/NewGameBtn';
import InfoTitle from '../../components/TicTacToe/InfoTitle';
import ResultsPanel from '../../components/TicTacToe/ResultsPanel';
import FieldCell from '../../components/TicTacToe/FieldCell';
import './style.scss';

export default class TicTacToe extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentPlayer: null,
      items: [],
      finishedGame: false,
      draw: false,
      totalSteps: 0,
      toggledStep: 0
    };
  }

  componentWillMount() {
    const { currentPlayer } = this.props;
    this.setState({
      currentPlayer,
      items: this.createInitalArr()
    });
  }

  startNewGame = () => {
    const { currentPlayer } = this.props;
    this.setState({
      currentPlayer,
      finishedGame: false,
      draw: false,
      items: this.createInitalArr()
    });
  }

  clickBack = () => {
    const { toggledStep, items } = this.state;
    const currentStep = toggledStep - 1;
    const arr = items;
    arr.forEach((item) => {
      if (item.step > currentStep) {
        item.visible = false;
      }
    });

    if (currentStep >= 0) {
      this.setState({
        toggledStep: currentStep,
        items: arr
      });
    }
  }

  clickForward = () => {
    const { toggledStep, items, totalSteps } = this.state;
    const currentStep = toggledStep + 1;
    const arr = items;
    arr.forEach((item) => {
      if (item.step <= currentStep) {
        item.visible = true;
      }
    });

    if (currentStep <= totalSteps) {
      this.setState({
        toggledStep: currentStep,
        items: arr
      });
    }
  }

  createInitalArr() {
    const { fieldLength } = this.props;
    const length = fieldLength;
    const arr = [];
    for (let i = 0; i < length; i++) {
      arr.push({
        key: i,
        value: null,
        visible: true,
        step: null,
        className: ''
      });
    }
    return arr;
  }

  handleClick(i) {
    const {
      items, currentPlayer, totalSteps, toggledStep
    } = this.state;
    const updatedItemsArr = items;
    const currentPlayerName = currentPlayer;
    updatedItemsArr[i] = {
      ...items[i],
      value: this.props[currentPlayerName].value,
      disabled: true,
      step: totalSteps + 1
    };

    this.setState({
      currentPlayer: (currentPlayerName === 'player1' ? 'player2' : 'player1'),
      items: updatedItemsArr,
      totalSteps: totalSteps + 1,
      toggledStep: toggledStep + 1
    });

    this.checkWinner(currentPlayerName);
  }

  checkWinner(currentPlayerName) {
    const { fieldLength } = this.props;
    const { items } = this.state;
    const checkedSign = this.props[currentPlayerName].value;
    const stringLength = Math.sqrt(fieldLength);
    const arr = items;
    const multiArr = [];
    const verticalArr = [];
    const diagonalTLRB = [];
    const diagonalLBRT = [];
    let horizontalArr = [];
    let winningArray = [];

    // Multidimentional array
    let multiStart = 0;
    let multiEnd = multiStart + stringLength;
    for (let i = 0; i < stringLength; i++) {
      multiArr.push(arr.slice(multiStart, multiEnd));
      multiStart = multiEnd;
      multiEnd = multiStart + stringLength;
    }

    // Horizontal
    horizontalArr = multiArr.slice();

    // Vertical
    for (let el = 0; el < stringLength; el++) {
      const subArr = [];
      for (let r = 0; r < stringLength; r++) {
        subArr.push(multiArr[r][el]);
      }
      verticalArr.push(subArr);
    }

    // Diagonal top-left/right-bottom
    for (let i = 0; i < stringLength; i++) {
      diagonalTLRB.push(multiArr[i][i]);
    }

    // Diagonal left-bottom/top-right
    let el = stringLength - 1;
    for (let i = 0; i < stringLength; i++) {
      diagonalLBRT.push(multiArr[i][el]);
      el--;
    }

    // Merge winning combination
    winningArray = [...horizontalArr, ...verticalArr, diagonalTLRB, diagonalLBRT];

    // Define winner
    let winnerCombination;
    for (let i = 0; i < winningArray.length; i++) {
      if (winningArray[i].every((val) => val.value === checkedSign)) {
        winnerCombination = winningArray[i];
      }
    }

    // Highlight win combination
    if (winnerCombination) {
      for (let i = 0; i < winnerCombination.length; i++) {
        winnerCombination.forEach((item) => {
          item.className = 'ttt-field__item--win';
        });
      }
      this.setState({
        finishedGame: true,
        currentPlayer: currentPlayerName
      });
      this.props.onIncreaseTotalSets();
      this.props.onIncreasePlayerWin(currentPlayerName);
    }

    // Draw
    if (arr.every((val) => val.value)) {
      this.setState({
        draw: true
      });
      this.props.onIncreaseTotalSets();
    }
  }

  render() {
    const {
      items, finishedGame, draw, currentPlayer
    } = this.state;
    const {
      setsPlayed, player1, player2
    } = this.props;

    return (
      <section className="ttt">
        <Helmet>
          <title>Tic Tac Toe</title>
        </Helmet>

        <ResultsPanel setsPlayed={setsPlayed} player1={player1} player2={player2} />

        {finishedGame
          && (
            <InfoTitle className="ttt-title--win">
              <span className="ttt-title__sign">{this.props[currentPlayer].value} </span>
              is winner. Our congatulations!
            </InfoTitle>
          )}
        {draw
          && (
            <InfoTitle className="ttt-title--draw">
              None of the players won.
            </InfoTitle>
          )}
        {!finishedGame && !draw ? (
          <InfoTitle>
            It is
            <span className="ttt-title__sign"> {this.props[currentPlayer].value} </span>
            turn.
          </InfoTitle>
        )
          : null
        }

        <div className="ttt-field-wrap">
          {finishedGame || draw ? (
            <NewGameBtn onStartNewGame={this.startNewGame} />
          ) : null
          }
          <div className="ttt-field">
            {items.map((item, index) => (
              <FieldCell
                key={item.key}
                disabled={item.disabled}
                visible={item.visible}
                value={item.value}
                className={item.className}
                onHandleClick={() => this.handleClick(index)}
              />
            ))}
          </div>

          <Nav onBack={this.clickBack} onForward={this.clickForward} />
        </div>
      </section>
    );
  }
}

TicTacToe.propTypes = {
  player1: PropTypes.object,
  player2: PropTypes.object,
  currentPlayer: PropTypes.string,
  fieldLength: PropTypes.number,
  setsPlayed: PropTypes.number,
  onIncreaseTotalSets: PropTypes.func,
  onIncreasePlayerWin: PropTypes.func
};
