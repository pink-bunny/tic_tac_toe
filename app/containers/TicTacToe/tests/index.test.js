import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Nav from 'components/TicTacToe/Nav';
import InfoTitle from 'components/TicTacToe/InfoTitle';
import ResultsPanel from 'components/TicTacToe/ResultsPanel';
import FieldCell from 'components/TicTacToe/FieldCell';
import TicTacToe from '../TicTacToe';

describe.only('<TicTacToe />', () => {
  let props;
  beforeEach(() => {
    props = {
      fieldLength: 9,
      winner: null,
      currentPlayer: 'player1',
      player1: {
        value: 'x',
        wins: 0
      },
      player2: {
        value: 'o',
        wins: 0
      },
      setsPlayed: 0,
    };
  });
  describe('<Nav />', () => {
    it('component exists inside tic tac toe', () => {
      const renderedComponent = shallow(<TicTacToe {...props} />);
      expect(renderedComponent.find(Nav).length).toBe(1);
    });

    it('component has orrect HTML', () => {
      const tree = renderer
        .create(<Nav onBack={() => {}} onForward={() => {}} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('should render a <InfoTitle />', () => {
    const renderedComponent = shallow(<TicTacToe {...props} />);
    expect(renderedComponent.find(InfoTitle).length).toBe(1);
  });

  it('should render a <ResultsPanel />', () => {
    const renderedComponent = shallow(<TicTacToe {...props} />);
    expect(renderedComponent.find(ResultsPanel).length).toBe(1);
  });

  it('should render a <FieldCell />', () => {
    const renderedComponent = shallow(<TicTacToe {...props} />);
    expect(renderedComponent.find(FieldCell).length).toBe(9);
  });
});
