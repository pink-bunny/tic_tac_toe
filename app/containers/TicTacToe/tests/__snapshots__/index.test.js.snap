// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicTacToe /> <Nav /> correct HTML 1`] = `
<div
  className="ttt-nav"
>
  <button
    className="ttt-nav__btn"
    onClick={[Function]}
    type="button"
  >
    Step Back
  </button>
  <button
    className="ttt-nav__btn"
    onClick={[Function]}
    type="button"
  >
    Step Forward
  </button>
</div>
`;
